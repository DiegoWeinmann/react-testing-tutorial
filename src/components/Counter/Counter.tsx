import React from 'react';

const Counter = () => {
  const [counter, setCounter] = React.useState(0);

  const increment = () => {
    setCounter(counter + 1);
  };

  const decrement = () => {
    if (counter > 0) {
      setCounter(counter - 1);
    }
  };

  return (
    <>
      <h1 data-testid='counter'>{counter}</h1>
      <button data-testid='button-up' onClick={increment}>
        Up
      </button>
      <button
        disabled={counter === 0 ? true : false}
        data-testid='button-down'
        onClick={decrement}
      >
        Down
      </button>
    </>
  );
};

export default Counter;
