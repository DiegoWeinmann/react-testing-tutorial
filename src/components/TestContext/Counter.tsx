import React from 'react';

import { CounterContext } from './TestContext';

const Counter = () => {
  const { counter, increment, decrement } = React.useContext(CounterContext);
  console.log(counter);
  return (
    <>
      <h1 data-testid='counter'>{counter}</h1>
      <button data-testid='button-up' onClick={increment}>
        Up
      </button>
      <button
        disabled={counter === 0 ? true : false}
        data-testid='button-down'
        onClick={decrement}
      >
        Down
      </button>
    </>
  );
};

export default Counter;
