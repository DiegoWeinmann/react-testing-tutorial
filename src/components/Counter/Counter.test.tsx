import React from 'react';
import { render, cleanup, fireEvent } from '@testing-library/react';
import Counter from './Counter';

const renderCounter = () => {
  return render(<Counter />);
};

afterEach(cleanup);

describe('<Counter/>', () => {
  it('matches the snapshot.', () => {
    const { asFragment } = renderCounter();
    expect(asFragment()).toMatchSnapshot();
  });

  it('should have a initial count of 0.', () => {
    const { getByTestId } = renderCounter();
    expect(getByTestId('counter')).toHaveTextContent('0');
  });

  it('has a Down button that is disabled.', () => {
    const { getByTestId } = renderCounter();
    expect(getByTestId('button-down')).toBeDisabled();
  });

  //  Events
  it('increments counter when clicking up button.', () => {
    const { getByTestId } = renderCounter();
    fireEvent.click(getByTestId('button-up'));
    expect(getByTestId('counter')).toHaveTextContent('1');
  });

  it('does not decrement counter when counter is 0.', () => {
    const { getByTestId } = renderCounter();
    fireEvent.click(getByTestId('button-down'));
    expect(getByTestId('counter')).toHaveTextContent('0');
  });

  it.only('decrements the counter when clicking down button.', () => {
    const { getByTestId } = renderCounter();
    fireEvent.click(getByTestId('button-up'));
    fireEvent.click(getByTestId('button-down'));
    expect(getByTestId('counter')).toHaveTextContent('0');
  });
});
