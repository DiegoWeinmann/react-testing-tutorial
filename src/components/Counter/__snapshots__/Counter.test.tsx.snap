// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Counter/> matches the snapshot. 1`] = `
<DocumentFragment>
  <h1
    data-testid="counter"
  >
    0
  </h1>
  <button
    data-testid="button-up"
  >
    Up
  </button>
  <button
    data-testid="button-down"
    disabled=""
  >
    Down
  </button>
</DocumentFragment>
`;
